conans/model/graph_lock.py

```diff
@@ -2,297 +2,465 @@
 
+    def save(self, path):
+        serialized_graph_str = self._dumps(path)
+        save(path, serialized_graph_str)
+
     @staticmethod
-    def loads(text, revisions_enabled):
+    def _loads(text, revisions_enabled):
         graph_json = json.loads(text)
         version = graph_json.get("version")
         if version:
-            version = Version(version)
-            if version < "0.2":
-                raise ConanException("This lockfile was created with a previous incompatible "
+            if version != LOCKFILE_VERSION:
+                raise ConanException("This lockfile was created with an incompatible "
                                      "version. Please regenerate the lockfile")
             # Do something with it, migrate, raise...
-        profile_host = graph_json.get("profile_host") or graph_json["profile"]
+        profile_host = graph_json.get("profile_host", None)
         profile_build = graph_json.get("profile_build", None)
         # FIXME: Reading private very ugly
-        profile_host, _ = _load_profile(profile_host, None, None)
+        if profile_host:
+            profile_host, _ = _load_profile(profile_host, None, None)
         if profile_build:
             profile_build, _ = _load_profile(profile_build, None, None)
-        graph_lock = GraphLock.from_dict(graph_json["graph_lock"])
-        graph_lock.revisions_enabled = revisions_enabled
+        graph_lock = GraphLock.deserialize(graph_json["graph_lock"], revisions_enabled)
         graph_lock_file = GraphLockFile(profile_host, profile_build, graph_lock)
         return graph_lock_file
 
-    def save(self, path):
-        if not path.endswith(".lock"):
-            path = os.path.join(path, LOCKFILE)
-        serialized_graph_str = self.dumps()
-        save(path, serialized_graph_str)
-
-    def dumps(self):
-        result = {"profile_host": self.profile_host.dumps(),
-                  "graph_lock": self.graph_lock.as_dict(),
+    def _dumps(self, path):
+        # Make the lockfile more reproducible by using a relative path in the node.path
+        # At the moment the node.path value is not really used, only its existence
+        path = os.path.dirname(path)
+        serial_lock = self._graph_lock.serialize()
+        for node in serial_lock["nodes"].values():
+            p = node.get("path")
+            if p:
+                try:  # In Windows with different drives D: C: this fails
+                    node["path"] = os.path.relpath(p, path)
+                except ValueError:
+                    pass
+        result = {"graph_lock": serial_lock,
                   "version": LOCKFILE_VERSION}
+        if self.profile_host:
+            result["profile_host"] = self._profile_host.dumps()
         if self.profile_build:
-            result["profile_build"] = self.profile_build.dumps()
+            result["profile_build"] = self._profile_build.dumps()
         return json.dumps(result, indent=True)
```

Comment:
`True` is probably not what we want here, it will be converted to `1`. So let's use an integer:

```suggestion
        return json.dumps(result, indent=4)
```